File: ml-service/scripts/exportData.js
import mongoose from "mongoose";
import dotenv from "dotenv";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

import connectDB from "../../backend/config/db.js";
import Stock from "../../backend/models/stock.js";
import MarketData from "../../backend/models/marketData.js";


const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({
  path: path.join(__dirname, ".env")
});
// CLI arg
const symbol = process.argv[2];

if (!symbol) {
  console.error("Usage: node exportData.js SYMBOL");
  process.exit(1);
}

// This resolves to:
// D:/TradeBuddy/ml-service/data/EURUSD=X
const symbolDir = path.resolve(
  __dirname,
  "../data",
  symbol
);

console.log("Export folder:", symbolDir);

// Intervals to export
const INTERVALS = [
  "5m",
  "15m",
  "30m",
  "1h",
  "1d"
];

(async function main() {
  try {
    await connectDB();

    const stockDoc = await Stock.findOne({
      symbol
    });

    if (!stockDoc) {
      console.error(
        "❌ Stock not found:",
        symbol
      );
      process.exit(1);
    }

    // Create:
    // ml-service/data/EURUSD=X
    if (!fs.existsSync(symbolDir)) {
      fs.mkdirSync(symbolDir, {
        recursive: true
      });
    }

    for (const interval of INTERVALS) {
      let data = await MarketData.find({
        stock: stockDoc._id,
        interval
      }).sort({ timestamp: 1 })
      .lean();

      if (!data.length) {
        console.log(
          `No data found for ${symbol} with interval ${interval}`
        );
        continue;
      }

      if (interval !== "1d") {
        data = data.filter(
          doc =>
            doc.movingAverage != null &&
            doc.ema != null &&
            doc.rsi != null &&
            doc.macd != null &&
            doc.bollingerUpper != null &&
            doc.bollingerLower != null
        );

        if (!data.length) {
          console.log(
            `No complete rows (indicators) for ${symbol} at ${interval}`
          );
          continue;
        }
      }

      const csvData = [
        [
          "timestampUTC",
          "localTime",
          "open",
          "high",
          "low",
          "close",
          "volume",
          "movingAverage",
          "ema",
          "rsi",
          "macd",
          "bollingerUpper",
          "bollingerLower"
        ],
        ...data.map(doc => {
          const utc =
            doc.timestamp.toISOString();

          const local = new Date(
            doc.timestamp
          ).toLocaleString(
            "en-IN",
            {
              timeZone: "Asia/Kolkata",
              hour12: false
            }
          );

          return [
            utc,
            local,
            doc.open,
            doc.high,
            doc.low,
            doc.close,
            doc.volume,
            doc.movingAverage,
            doc.ema,
            doc.rsi,
            doc.macd,
            doc.bollingerUpper,
            doc.bollingerLower
          ];
        })
      ]
        .map(row => row.join(","))
        .join("\n");

      const outPath = path.join(
        symbolDir,
        `${symbol}_${interval}.csv`
      );

      fs.writeFileSync(
        outPath,
        csvData
      );

      console.log(
        `✅ Exported ${data.length} rows to ${outPath}`
      );
    }

    await mongoose.connection.close();

    console.log(
      "✅ Export completed successfully"
    );

    process.exit(0);

  } catch (err) {

    console.error(
      "❌ Error exporting data:",
      err
    );

    process.exit(1);
  }
})();

File: ml-service/database/export_data.py
import os
import pandas as pd
from datetime import timezone
from zoneinfo import ZoneInfo

from database.mongodb import stocks, market_data


INTERVALS = [
    "5m",
    "15m",
    "30m",
    "1h",
    "1d"
]


def export_symbol(symbol):

    stock = stocks.find_one({
        "symbol": symbol
    })

    if not stock:
        raise Exception(
            f"Stock not found: {symbol}"
        )


    symbol_dir = os.path.join(
        "data",
        symbol
    )

    os.makedirs(
        symbol_dir,
        exist_ok=True
    )


    for interval in INTERVALS:

        cursor = market_data.find(
            {
                "stock": stock["_id"],
                "interval": interval
            }
        ).sort(
            "timestamp",
            1
        )


        rows = list(cursor)


        if not rows:
            print(
                f"No data for {symbol} {interval}"
            )
            continue


        processed = []


        for doc in rows:

            if interval != "1d":

                indicators = [
                    doc.get("movingAverage"),
                    doc.get("ema"),
                    doc.get("rsi"),
                    doc.get("macd"),
                    doc.get("bollingerUpper"),
                    doc.get("bollingerLower")
                ]


                if any(x is None for x in indicators):
                    continue


            utc_time = doc["timestamp"]


            local_time = (
                utc_time
                .astimezone(
                    ZoneInfo("Asia/Kolkata")
                )
                .strftime(
                    "%d/%m/%Y, %H:%M:%S"
                )
            )


            processed.append({
                "timestampUTC":
                    utc_time.isoformat(),

                "localTime":
                    local_time,

                "open":
                    doc["open"],

                "high":
                    doc["high"],

                "low":
                    doc["low"],

                "close":
                    doc["close"],

                "volume":
                    doc["volume"],

                "movingAverage":
                    doc.get("movingAverage"),

                "ema":
                    doc.get("ema"),

                "rsi":
                    doc.get("rsi"),

                "macd":
                    doc.get("macd"),

                "bollingerUpper":
                    doc.get("bollingerUpper"),

                "bollingerLower":
                    doc.get("bollingerLower")
            })


        if not processed:
            print(
                f"No complete rows for {interval}"
            )
            continue


        df = pd.DataFrame(processed)


        output_path = os.path.join(
            symbol_dir,
            f"{symbol}_{interval}.csv"
        )


        df.to_csv(
            output_path,
            index=False
        )


        print(
            f"Exported {len(df)} rows to {output_path}"
        )


    print(
        "Export completed successfully"
    )